### DISCUSSION TITLE
concatenate arrays of records without getting Unions

### DISCUSSION THREAD

**Discussion by ivirshup** [scikit-hep/awkward#1647] [Q&A]
I would like to concatenate arrays of records without getting unions back. Here's a quick illustrative example of a use case:

```python
import awkward._v2 as ak

awk_a = ak.Array([
    {"a": [1, 2, 3], "b": [1, 2]},
    {"a": [4, 5], "b": [3, 4]},
    {"a": [6], "b": [5]},
])


awk_b = ak.Array([
    {"a": [1, 2, 3]},
    {"a": [4, 5]},
    {"a": [6]},
])

ak.concatenate([awk_a, awk_b])
```

I would like to end up with a non-Union result here. E.g. something like:

```python
# '6 * {a: var * int64, b: var * int64}'
ak.Array([
    {"a": [1, 2, 3], "b": [1, 2]},
    {"a": [4, 5], "b": [3, 4]},
    {"a": [6], "b": [5]},
    {"a": [1, 2, 3], "b": []},
    {"a": [4, 5], "b": []},
    {"a": [6], "b": []},
])

# '6 * {a: var * int64, b: option[var * int64]}'
ak.Array([
    {"a": [1, 2, 3], "b": [1, 2]},
    {"a": [4, 5], "b": [3, 4]},
    {"a": [6], "b": [5]},
    {"a": [1, 2, 3]},
    {"a": [4, 5]},
    {"a": [6]},
])
```

I would ideally like to be able to do this at concatenation time (and even be able to choose just to concatenate the intersection of keys). But, being able to convert from a union to an optional type would also work.

I'm having some trouble finding docs for this, how would I go about doing this?

**Comment by agoose77** [comment]
I'd suggest transforming `awk_b` such that it has the necessary field. You could make use of the `ak.Array.__setitem__` feature here, which allows you to add fields to an existing array (advanced users note that you can modify arrays, but not layouts, so this feature replaces the existing layout with a new one). The nice benefit of using `__setitem__` (or indeed `ak.with_field`) is that it broadcasts, so by doing

```python
awk_b['b'] = None
```

you end up with
```python
>>> awk_b.type.show()
3 * {
    a: var * int64,
    b: ?unknown
}
```

Upon concatenation, this then becomes:
```python
>>> ak.concatenate([awk_a, awk_b]).type.show()
6 * {
    a: var * int64,
    b: option[var * int64]
}
```

Awkward is still creating a union for the different layout nodes under the hood; `unknown` translates to `EmptyArray` in Awkward's layout node structure, whilst `var * int64` is a list-type over a NumPy array. Yet, Awkward has rules about which types can be merged, and which require a union, and `EmptyArray` can always be merged.

**Reply by ivirshup** [reply]
Thanks for the info!

Can I set field `"b"` not to be optional, but to be empty arrays of a fixed type? Maybe something like:

```python
ak.with_field(awk_b, ListType(NumpyType('int64')), "b")
```

---------

Are there function for operations on types? E.g. Find a common subset between two array types?

**Reply by agoose77** [reply]
At present the easiest ways to do this are to take an existing array and e.g. slice it to zero length
```python3
awk_b['b'] = awk_b['a'][..., :0]
```

It really depends upon what type field `b` is supposed to be, and how it corresponds to the other fields.

The mechanism for building layouts from type information (besides directly instantiating `ak._v2.contents.Content` subclasses) is to use `ak._v2.from_buffers` with `form` information. There is no mapping from type objects to forms, so you'd need to create the form by hand, or from an existing array.

There are no type operations per-se. We do have routines that merge different types, so we have logic that can determine if types are mergeable. We don't do anything with subsets of fields, IIRC, though.

## Advanced
You can manually build a `ak._v2.contents.Content` layout yourself if you know what you need. For example, the layout of `awk_b` after applying the above is:
```python3
<RecordArray is_tuple='false' len='3'>
    <content index='0' field='a'>
        <ListOffsetArray len='3'>
            <offsets><Index dtype='int64' len='4'>
                [0 3 5 6]
            </Index></offsets>
            <content><NumpyArray dtype='int64' len='6'>[1 2 3 4 5 6]</NumpyArray></content>
        </ListOffsetArray>
    </content>
    <content index='1' field='b'>
        <ListOffsetArray len='3'>
            <offsets><Index dtype='int64' len='4'>
                [0 0 0 0]
            </Index></offsets>
            <content><NumpyArray dtype='int64' len='0'>[]</NumpyArray></content>
        </ListOffsetArray>
    </content>
</RecordArray>
```

Similarly, you can use `ak._v2.from_buffers` to build an array from a form and a flat list of buffers.

**Reply by ivirshup** [reply]
From the other conversation, do you think `ak.transform` could be a good way to implement this?

That way I could recurse down a set of arrays and "broadcast" them to an inner or outer union before concatenation.

**Reply by jpivarski** [reply]
The `__setitem__` that @agoose77 suggested,

> At present the easiest ways to do this are to take an existing array and e.g. slice it to zero length
> 
> ```python
> awk_b['b'] = awk_b['a'][..., :0]
> ```

does use broadcasting to create the new field. (It's implemented with the same backend as `ak.transform`.) Since this high-level mechanism exists to do it, this is definitely simpler than rolling your own.

**Reply by ivirshup** [reply]
Just trying this out now, and running into an issue. I can get the expected result from:

```python
awk_a = ak.Array([
    {"a": [1, 2, 3], "b": [1, 2]},
    {"a": [4, 5], "b": [3, 4]},
    {"a": [6], "b": [5]},
])


awk_b = ak.Array([
    {"a": [1, 2, 3]},
    {"a": [4, 5]},
    {"a": [6]},
])

ak.concatenate([awk_a, ak.with_field(awk_b, None, "b")]).type.show()
```

```
6 * {
    a: var * int64,
    b: option[var * int64]
}
```

However, I get a `union` from:

```python
awk_c = ak.Array([
    [{"a": 1, "b": "foo"}],
    [{"a": 2, "b": "bar"}, {"a": 3, "b": "baz"}]
])

awk_d = ak.Array([
    [{"a": 4}, {"a": 5}],
    [{"a": 6}],
    [{"a": 7}],
])

ak.concatenate([awk_c, ak.with_field(awk_d, None, "b")]).type.show()
```

```
5 * union[
    var * {
        a: int64,
        b: string
    },
    var * {
        a: int64,
        b: ?unknown
    }
]
```

Is this expected?

**Reply by ivirshup** [reply]
As an aside, is there a way to do something like:

```python
awk_c.astype(ak.types.from_datashape("2 * var * {a: int64, b: ?string}"))
```

If not, is there a fundamental barrier to something like this being possible?

**Reply by jpivarski** [reply]
> Is this expected?

I think you found a bug: we want "string" and "option[unknown]" to unify to "option[string]", and it's not recognizing that. It should first notice that one of the two is "option" and make the result type "option", then unify "string" (or "X") with "unknown" to get "string" ("X"), since "unknown" is an identity in type-unification. I _think_ this is supposed to happen somewhere in [simplify_uniontype](https://github.com/scikit-hep/awkward/blob/1bcfd70629aec4c1ef26bbf929d9183c28dbe491/src/awkward/_v2/contents/unionarray.py#L471-L661). It's interesting that it got it right when "X" is a primitive type, int64, but not when it's a list type. This is something we should promote into an issue.

As for casting an entire array as an entire type, that sounds like a good function to have, though it would take some thinking to do it right. You'd have to recurse down the array tree and the type tree, modifying array elements based on what we see in the type, and it wouldn't be one-to-one because high-level types are simpler than arrays. So, it's not a fundamental barrier, but it will be quicker for us to fix the bug in type-unification, above.
